rofl-appd/src/routes/tx.rs

```diff
@@ -53,21 +54,43 @@ pub enum Transaction {
     /// Ethereum transaction.
     #[serde(rename = "eth")]
     Eth {
-        gas_limit: u64,
-        #[serde_as(as = "serde_with::hex::Hex")]
+        gas_limit: QuantityU64,
+        #[serde(deserialize_with = "deserialize_hex_bytes")]
         to: Vec<u8>,
         value: TransactionValue,
-        #[serde_as(as = "serde_with::hex::Hex")]
+        #[serde(deserialize_with = "deserialize_hex_bytes")]
         data: Vec<u8>,
     },
 }
 
+/// A quantity that can be provided either as a JSON number or as a string
+/// (decimal or 0x-prefixed hex) and normalized into a `u64`.
+#[derive(Clone, Debug, serde::Deserialize)]
+#[serde(untagged)]
+pub enum QuantityU64 {
+    Number(u64),
+    String(String),
+}
+
+impl QuantityU64 {
+    fn into_u64(self, field: &'static str) -> Result<u64, String> {
+        match self {
+            QuantityU64::Number(n) => Ok(n),
+            QuantityU64::String(s) => parse_u64_string(s, field),
+        }
+    }
+}
+
 /// Value representation that accepts either a string (decimal or 0x hex) or a JSON number.
+///
+/// Note: The `Number` variant uses `u64` because serde_json does not support `u128` in untagged
+/// enums. For values larger than `u64::MAX`, use the `String` variant with a decimal or hex
+/// representation.
 #[derive(Clone, Debug, serde::Deserialize)]
 #[serde(untagged)]
 pub enum TransactionValue {
     String(String),
-    Number(u128),
+    Number(u64),
```

Comment:
Transaction value and gas price must be at least `u128`. You can't fit the circulating supply of the tokens into 64 bits (in wei). e.g. 20B ROSE requires 95 bits.

Reply:
@matevz String variant already supports values up to `u256`, which handles 20B ROSE in wei, added a test to verify. Number(u64) variant is limited because `serde_json` can't deserialize JSON integers > `u64` into Rust `u128` in untagged enums, clients should use string format for large values (which is also safer since JavaScript's `Number.MAX_SAFE_INTEGER` is only 2^53-1)